binding/tfe_utils.cc

```diff
@@ -270,12 +271,11 @@ void AssignOpAttr(napi_env env, TFE_Op *tfe_op, napi_value attr_value) {
     case TF_ATTR_STRING: {
       // NOTE: String attribute values do not have to be utf8 encoded strings
       // (could be arbitrary byte sequences).
-      char value[NAPI_STRING_SIZE];
-      nstatus = napi_get_value_string_utf8(env, js_value, value,
-                                           NAPI_STRING_SIZE, nullptr);
-      ENSURE_NAPI_OK(env, nstatus);
+      std::string *str_value = scoped_strings.GetString(env, js_value);
```

Comment:
Can we avoid the `TF_ScopedString` class altogether and just use the length returned from [`napi_get_value_string_utf8`](https://nodejs.org/api/n-api.html#n_api_napi_create_string_utf8)? Doing so will also avoid a string copy, I think.

Something like:

```c++
case TF_ATTR_STRING: {
  char value[NAPI_STRING_SIZE];
  size_t size;
  nstatus = napi_get_value_string_utf8(env, js_value, value, NAPI_STRING_SIZE, &size);
  ENSURE_NAPI_OK(env, nstatus);
  // May also want to compare size with NAPI_STRING_SIZE to detect truncation?
  TFE_OpSetAttrString(tfe_op, attr_name, value, size);
  break;
}
```

Reply:
I tried that initially - but the string is not copied. Since StringPiece just refers to a pointer, multiple Op attributes will just point to the last string value - hence the new ScopedString class.

Reply:
Ah, hmm....that is unfortunate. Perhaps `TFE_OpSetAttrString` should not hold on to memory after it returns (i.e., it should make a copy). Though, this aspect didn't change recently, it was still using a `StringPiece` before https://github.com/tensorflow/tensorflow/commit/2b0805301e4531dd7c2ed677d932f6408675460e

Does that mean that we always had a bug in handling string valued attributes? (i.e., this is not new to 1.10.0?)

Reply:
Previous builds from the r1.9 branch worked just fine - I only hit this problem when running builds from the r1.10 branch.

It looks like RC2 went out on July 2nd - https://github.com/tensorflow/tensorflow/releases/tag/v1.9.0-rc2 - but our builds were from the v1.9 branch - so maybe we missed this. I can verify tomorrow.

Reply:
Also, I might be mis-reading this - but this commit changed to StringPiece: https://github.com/tensorflow/tensorflow/commit/2b0805301e4531dd7c2ed677d932f6408675460e :

Old:

```c++
void TFE_OpSetAttrString(TFE_Op* op, const char* attr_name, const char* value) {
  op->operation.MutableAttrs()->Set(attr_name, value);
```

New
```c++
void TFE_OpSetAttrString(TFE_Op* op, const char* attr_name, const void* value,
                         size_t length) {
  op->operation.MutableAttrs()->Set(
      attr_name,
      tensorflow::StringPiece(static_cast<const char*>(value), length));
```

Reply:
Hmmm..you might be right.

I thought that the commit just made the `StringPiece` conversion explicit as when `value` is a `const char*` in:

```c++
op->operation.MutableAttrs()->Set(attr_name, value);
```

it would still resolve to: https://github.com/tensorflow/tensorflow/blob/c2399dbe405eb8cbdbe0451336f46c6085d9a3c5/tensorflow/core/common_runtime/eager/attr_builder.h#L146

but I might be mistaken and it resolved to:
https://github.com/tensorflow/tensorflow/blob/c2399dbe405eb8cbdbe0451336f46c6085d9a3c5/tensorflow/core/common_runtime/eager/attr_builder.h#L97

instead. If so, then this seems like a bug in that commit that I'd want to fix :)
(And perhaps your change should go ahead in the meantime with a TODO to make it simpler when 1.11 is released?)

Reply:
Sounds good - I'm happy to submit a fix. Should I file an issue/bug for this? 

I pushed a commit that contains TODOs for me to cleanup on 1.11.